src/main/java/build/buildfarm/worker/Worker.java

```diff
@@ -205,9 +207,20 @@ private boolean linkInputs(Path execDir, Digest inputRoot, Map<Digest, Directory
     for (FileNode fileNode : directory.getFilesList()) {
       Digest digest = fileNode.getDigest();
       Path fileCachePath = cacheDir.resolve(String.format("%s_%d", digest.getHash(), digest.getSizeBytes()));
-      OutputStream outputStream = Files.newOutputStream(fileCachePath);
-      instance.getBlob(digest).writeTo(outputStream);
-      outputStream.close();
+      boolean createFile = !Files.exists(fileCachePath);
+      if (createFile) {
+        OutputStream outputStream = Files.newOutputStream(fileCachePath);
+        instance.getBlob(digest).writeTo(outputStream);
+        outputStream.close();
+      }
+      if (createFile || Files.isExecutable(fileCachePath) != fileNode.getIsExecutable()) {
+        ImmutableSet.Builder<PosixFilePermission> perms = new ImmutableSet.Builder<PosixFilePermission>()
+          .add(PosixFilePermission.OWNER_READ);
+        if (fileNode.getIsExecutable()) {
```

Comment:
I'll say that there are other areas of the worker's operation that will have to be adjusted if the cache is shared amongst many content manipulators - the individual files in the cache will need to be guarded for example.  I would be happy to document that for this worker implementation, the cache needs to be exclusive (for now), where do you see that going?  README.md in the root?